### DISCUSSION TITLE
Commit order enforcement when using connection.execute() within after_insert event handler

### DISCUSSION THREAD

**Discussion by jdimmerman** [sqlalchemy/sqlalchemy#11928] [Usage Questions] [answered]
Hello - I'm trying to confirm some behavior I'm seeing as its critical to our system:

We are performing a large system refactor, part of which requires strangling from one legacy model to a new model. In order to keep the two models in sync, we are leveraging after_insert and after_update. We are using the ORM style syntax. Example:

1. Update record in legacy model
2. after_update fires, though transaction isn't yet committed
3. within after_update, perform connection.execute() to sync models

Our backend is psql and #3 generally looks like:
```
    connection.execute(
        update(NewModel)
        .where(NewModel.legacy_model_id == LegacyModel.id)
        .values({
            "some_field": LegacyModel.mapped_legacy_some_field,
            ...
        })
    )
```

We have been leveraging this at scale for a few months without issue, but I realized that there may be a large gap as I learned recently that after_insert is called prior to the transaction being committed: In the after_update handler, does `connection.execute()` operate within the same transaction or is order somehow otherwise guaranteed?

Thanks!

**Comment by zzzeek** [comment, marked as answer]
The [after_insert](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.after_insert) and [after_update](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.after_update) events are both invoked within the [ORM flush process](https://docs.sqlalchemy.org/en/20/tutorial/orm_data_manipulation.html#flushing), which itself is always contained within the transaction used by the `Session`, all of which is before the database transaction is committed.   The transaction begin/commit is always the outermost phase of all SQL that's emitted within the scope of a `Session`.

**Reply by jdimmerman** [reply]
Thanks @zzzeek! And to confirm - `connection.execute()` executes commands within that same `Session`?

**Reply by zzzeek** [reply]
that connection is the one that flush() is using, so your SQL are right inline on the same transaction sure

**Reply by jdimmerman** [reply]
Amazing - thank you @zzzeek. This puts my mind at ease
